File: mycollab-web/src/main/java/com/esofthead/mycollab/module/user/accountsettings/view/UserAccountController.java
package com.esofthead.mycollab.module.user.accountsettings.view;

import com.esofthead.mycollab.common.localization.GenericI18Enum;
import com.esofthead.mycollab.core.arguments.NumberSearchField;
import com.esofthead.mycollab.core.arguments.SearchField;
import com.esofthead.mycollab.core.utils.LocalizationHelper;
import com.esofthead.mycollab.module.user.accountsettings.billing.view.BillingSummaryPresenter;
import com.esofthead.mycollab.module.user.accountsettings.profile.view.ProfilePresenter;
import com.esofthead.mycollab.module.user.accountsettings.team.view.UserPermissionManagementPresenter;
import com.esofthead.mycollab.module.user.accountsettings.view.events.AccountBillingEvent;
import com.esofthead.mycollab.module.user.accountsettings.view.events.ProfileEvent;
import com.esofthead.mycollab.module.user.accountsettings.view.parameters.ProfileScreenData;
import com.esofthead.mycollab.module.user.accountsettings.view.parameters.RoleScreenData;
import com.esofthead.mycollab.module.user.accountsettings.view.parameters.UserScreenData;
import com.esofthead.mycollab.module.user.domain.Role;
import com.esofthead.mycollab.module.user.domain.SimpleRole;
import com.esofthead.mycollab.module.user.domain.SimpleUser;
import com.esofthead.mycollab.module.user.domain.criteria.RoleSearchCriteria;
import com.esofthead.mycollab.module.user.domain.criteria.UserSearchCriteria;
import com.esofthead.mycollab.module.user.events.RoleEvent;
import com.esofthead.mycollab.module.user.events.UserEvent;
import com.esofthead.mycollab.module.user.service.RoleService;
import com.esofthead.mycollab.vaadin.events.ApplicationEvent;
import com.esofthead.mycollab.vaadin.events.ApplicationEventListener;
import com.esofthead.mycollab.vaadin.events.EventBus;
import com.esofthead.mycollab.vaadin.mvp.IController;
import com.esofthead.mycollab.vaadin.mvp.PresenterResolver;
import com.esofthead.mycollab.web.AppContext;
import com.vaadin.ui.Window;

public class UserAccountController implements IController {
	private static final long serialVersionUID = 1L;

	private AccountModule container;

	public UserAccountController(AccountModule container) {
		this.container = container;

		bindProfileEvents();
		bindBillingEvents();
		bindRoleEvents();
		bindUserEvents();
	}

	private void bindBillingEvents() {
		EventBus.getInstance()
				.addListener(
						new ApplicationEventListener<AccountBillingEvent.GotoSummary>() {
							private static final long serialVersionUID = 1L;

							@Override
							public Class<? extends ApplicationEvent> getEventType() {
								return AccountBillingEvent.GotoSummary.class;
							}

							@Override
							public void handle(
									AccountBillingEvent.GotoSummary event) {
								BillingSummaryPresenter presenter = PresenterResolver
										.getPresenter(BillingSummaryPresenter.class);
								presenter.go(container, null);
							}
						});
	}

	private void bindProfileEvents() {
		EventBus.getInstance().addListener(
				new ApplicationEventListener<ProfileEvent.GotoUploadPhoto>() {
					private static final long serialVersionUID = 1L;

					@Override
					public Class<? extends ApplicationEvent> getEventType() {
						return ProfileEvent.GotoUploadPhoto.class;
					}

					@Override
					public void handle(ProfileEvent.GotoUploadPhoto event) {
						ProfilePresenter presenter = PresenterResolver
								.getPresenter(ProfilePresenter.class);
						presenter.go(container,
								new ProfileScreenData.UploadPhoto(
										(byte[]) event.getData()));
					}
				});

		EventBus.getInstance().addListener(
				new ApplicationEventListener<ProfileEvent.GotoProfileView>() {
					private static final long serialVersionUID = 1L;

					@Override
					public Class<? extends ApplicationEvent> getEventType() {
						return ProfileEvent.GotoProfileView.class;
					}

					@Override
					public void handle(ProfileEvent.GotoProfileView event) {
						ProfilePresenter presenter = PresenterResolver
								.getPresenter(ProfilePresenter.class);
						presenter.go(container, null);
					}
				});
	}

	private void bindUserEvents() {
		EventBus.getInstance().addListener(
				new ApplicationEventListener<UserEvent.GotoAdd>() {
					private static final long serialVersionUID = 1L;

					@Override
					public Class<? extends ApplicationEvent> getEventType() {
						return UserEvent.GotoAdd.class;
					}

					@Override
					public void handle(UserEvent.GotoAdd event) {
						UserPermissionManagementPresenter presenter = PresenterResolver
								.getPresenter(UserPermissionManagementPresenter.class);
						presenter.go(container, new UserScreenData.Add(
								new SimpleUser()));
					}
				});

		EventBus.getInstance().addListener(
				new ApplicationEventListener<UserEvent.GotoEdit>() {
					private static final long serialVersionUID = 1L;

					@Override
					public Class<? extends ApplicationEvent> getEventType() {
						return UserEvent.GotoEdit.class;
					}

					@Override
					public void handle(UserEvent.GotoEdit event) {
						UserPermissionManagementPresenter presenter = PresenterResolver
								.getPresenter(UserPermissionManagementPresenter.class);

						SimpleUser user = (SimpleUser) event.getData();
						presenter.go(container, new UserScreenData.Edit(user));
					}
				});

		EventBus.getInstance().addListener(
				new ApplicationEventListener<UserEvent.GotoRead>() {
					private static final long serialVersionUID = 1L;

					@Override
					public Class<? extends ApplicationEvent> getEventType() {
						return UserEvent.GotoRead.class;
					}

					@Override
					public void handle(UserEvent.GotoRead event) {
						UserPermissionManagementPresenter presenter = PresenterResolver
								.getPresenter(UserPermissionManagementPresenter.class);
						presenter.go(container, new UserScreenData.Read(
								(String) event.getData()));
					}
				});

		EventBus.getInstance().addListener(
				new ApplicationEventListener<UserEvent.GotoList>() {
					private static final long serialVersionUID = 1L;

					@Override
					public Class<? extends ApplicationEvent> getEventType() {
						return UserEvent.GotoList.class;
					}

					@Override
					public void handle(UserEvent.GotoList event) {
						UserPermissionManagementPresenter presenter = PresenterResolver
								.getPresenter(UserPermissionManagementPresenter.class);

						UserSearchCriteria criteria = new UserSearchCriteria();
						criteria.setSaccountid(new NumberSearchField(
								SearchField.AND, AppContext.getAccountId()));
						presenter.go(container, new UserScreenData.Search(
								criteria));
					}
				});
	}

	private void bindRoleEvents() {
		EventBus.getInstance().addListener(
				new ApplicationEventListener<RoleEvent.GotoAdd>() {
					private static final long serialVersionUID = 1L;

					@Override
					public Class<? extends ApplicationEvent> getEventType() {
						return RoleEvent.GotoAdd.class;
					}

					@Override
					public void handle(RoleEvent.GotoAdd event) {
						UserPermissionManagementPresenter presenter = PresenterResolver
								.getPresenter(UserPermissionManagementPresenter.class);
						presenter.go(container, new RoleScreenData.Add(
								new Role()));
					}
				});

		EventBus.getInstance().addListener(
				new ApplicationEventListener<RoleEvent.GotoEdit>() {
					private static final long serialVersionUID = 1L;

					@Override
					public Class<? extends ApplicationEvent> getEventType() {
						return RoleEvent.GotoEdit.class;
					}

					@Override
					public void handle(RoleEvent.GotoEdit event) {
						UserPermissionManagementPresenter presenter = PresenterResolver
								.getPresenter(UserPermissionManagementPresenter.class);

						Role role = (Role) event.getData();
						presenter.go(container, new RoleScreenData.Edit(role));
					}
				});

		EventBus.getInstance().addListener(
				new ApplicationEventListener<RoleEvent.GotoRead>() {
					private static final long serialVersionUID = 1L;

					@Override
					public Class<? extends ApplicationEvent> getEventType() {
						return RoleEvent.GotoRead.class;
					}

					@Override
					public void handle(RoleEvent.GotoRead event) {
						UserPermissionManagementPresenter presenter = PresenterResolver
								.getPresenter(UserPermissionManagementPresenter.class);
						if (event.getData() instanceof SimpleRole) {
							presenter.go(container, new RoleScreenData.Read(
									(Role) event.getData()));
						} else if (event.getData() instanceof Integer) {
							RoleService roleService = AppContext
									.getSpringBean(RoleService.class);
							SimpleRole role = roleService.findById(
									(Integer) event.getData(),
									AppContext.getAccountId());
							if (role == null) {
								AppContext
										.getApplication()
										.getMainWindow()
										.showNotification(
												LocalizationHelper
														.getMessage(GenericI18Enum.INFORMATION_WINDOW_TITLE),
												LocalizationHelper
														.getMessage(GenericI18Enum.INFORMATION_RECORD_IS_NOT_EXISTED_MESSAGE),
												Window.Notification.TYPE_HUMANIZED_MESSAGE);
							} else {
								presenter.go(container,
										new RoleScreenData.Read(role));
							}
						}
					}
				});

		EventBus.getInstance().addListener(
				new ApplicationEventListener<RoleEvent.GotoList>() {
					private static final long serialVersionUID = 1L;

					@Override
					public Class<? extends ApplicationEvent> getEventType() {
						return RoleEvent.GotoList.class;
					}

					@Override
					public void handle(RoleEvent.GotoList event) {
						UserPermissionManagementPresenter presenter = PresenterResolver
								.getPresenter(UserPermissionManagementPresenter.class);

						RoleSearchCriteria criteria = new RoleSearchCriteria();
						criteria.setSaccountid(new NumberSearchField(
								SearchField.AND, AppContext.getAccountId()));
						presenter.go(container, new RoleScreenData.Search(
								criteria));
					}
				});
	}
}

File: mycollab-web/src/main/scala/com/mycollab/module/user/accountsettings/view/UserAccountController.scala
package com.mycollab.premium.module.user.accountsettings.view

import com.google.common.eventbus.Subscribe
import com.mycollab.configuration.SiteConfiguration
import com.mycollab.db.arguments.{NumberSearchField, SetSearchField}
import com.mycollab.eventmanager.ApplicationEventListener
import com.mycollab.module.billing.RegisterStatusConstants
import com.mycollab.module.user.accountsettings.billing.view.IBillingPresenter
import com.mycollab.module.user.accountsettings.customize.view.AccountSettingPresenter
import com.mycollab.module.user.accountsettings.profile.view.ProfilePresenter
import com.mycollab.module.user.accountsettings.setup.view.SetupPresenter
import com.mycollab.module.user.accountsettings.team.view.UserPermissionManagementPresenter
import com.mycollab.module.user.accountsettings.view.AccountModule
import com.mycollab.module.user.accountsettings.view.events.{AccountBillingEvent, ProfileEvent, SettingEvent, SetupEvent}
import com.mycollab.module.user.accountsettings.view.parameters.SettingExtScreenData.{GeneralSetting, ThemeCustomize}
import com.mycollab.module.user.accountsettings.view.parameters.{BillingScreenData, RoleScreenData, UserScreenData}
import com.mycollab.module.user.domain.criteria.{RoleSearchCriteria, UserSearchCriteria}
import com.mycollab.module.user.domain.{Role, SimpleUser}
import com.mycollab.module.user.events.{RoleEvent, UserEvent}
import com.mycollab.vaadin.AppUI.getAccountId
import com.mycollab.vaadin.{AppUI, UserUIContext}
import com.mycollab.vaadin.mvp.{AbstractController, PresenterResolver}

/**
  * @author MyCollab Ltd
  * @since 5.0.9
  */
class UserAccountController(container: AccountModule) extends AbstractController {
  bindProfileEvents()
  bindBillingEvents()
  bindRoleEvents()
  bindUserEvents()
  bingSettingEvents()

  if (!SiteConfiguration.isDemandEdition) {
    bindSetupEvents()
  }

  private def bindBillingEvents(): Unit = {
    this.register(new ApplicationEventListener[AccountBillingEvent.CancelAccount]() {
      @Subscribe def handle(event: AccountBillingEvent.CancelAccount) {
        val presenter = PresenterResolver.getPresenter(classOf[IBillingPresenter])
        presenter.go(container, new BillingScreenData.CancelAccount)
      }
    })
    this.register(new ApplicationEventListener[AccountBillingEvent.GotoSummary]() {
      @Subscribe def handle(event: AccountBillingEvent.GotoSummary) {
        val presenter = PresenterResolver.getPresenter(classOf[IBillingPresenter])
        presenter.go(container, new BillingScreenData.BillingSummary)
      }
    })
    this.register(new ApplicationEventListener[AccountBillingEvent.GotoHistory]() {
      @Subscribe def handle(event: AccountBillingEvent.GotoHistory) {
        val presenter = PresenterResolver.getPresenter(classOf[IBillingPresenter])
        presenter.go(container, new BillingScreenData.BillingHistory)
      }
    })
  }

  private def bindProfileEvents(): Unit = {
    this.register(new ApplicationEventListener[ProfileEvent.GotoProfileView]() {
      @Subscribe def handle(event: ProfileEvent.GotoProfileView) {
        val presenter = PresenterResolver.getPresenter(classOf[ProfilePresenter])
        presenter.go(container, null)
      }
    })
  }

  private def bindUserEvents(): Unit = {
    this.register(new ApplicationEventListener[UserEvent.GotoAdd]() {
      @Subscribe def handle(event: UserEvent.GotoAdd) {
        val presenter = PresenterResolver.getPresenter(classOf[UserPermissionManagementPresenter])
        presenter.go(container, new UserScreenData.Add(new SimpleUser))
      }
    })
    this.register(new ApplicationEventListener[UserEvent.GotoEdit]() {
      @Subscribe def handle(event: UserEvent.GotoEdit) {
        val presenter = PresenterResolver.getPresenter(classOf[UserPermissionManagementPresenter])
        val user = event.getData.asInstanceOf[SimpleUser]
        presenter.go(container, new UserScreenData.Edit(user))
      }
    })
    this.register(new ApplicationEventListener[UserEvent.GotoRead]() {
      @Subscribe def handle(event: UserEvent.GotoRead) {
        val presenter = PresenterResolver.getPresenter(classOf[UserPermissionManagementPresenter])
        presenter.go(container, new UserScreenData.Read(event.getData.asInstanceOf[String]))
      }
    })
    this.register(new ApplicationEventListener[UserEvent.GotoList]() {
      @Subscribe def handle(event: UserEvent.GotoList) {
        val presenter = PresenterResolver.getPresenter(classOf[UserPermissionManagementPresenter])
        val criteria = new UserSearchCriteria
        criteria.setSaccountid(new NumberSearchField(AppUI.getAccountId))
        criteria.setRegisterStatuses(new SetSearchField[String](RegisterStatusConstants.ACTIVE, RegisterStatusConstants.NOT_LOG_IN_YET))
        presenter.go(container, new UserScreenData.Search(criteria))
      }
    })
  }

  private def bindRoleEvents(): Unit = {
    this.register(new ApplicationEventListener[RoleEvent.GotoAdd]() {
      @Subscribe def handle(event: RoleEvent.GotoAdd) {
        val presenter = PresenterResolver.getPresenter(classOf[UserPermissionManagementPresenter])
        presenter.go(container, new RoleScreenData.Add(new Role))
      }
    })
    this.register(new ApplicationEventListener[RoleEvent.GotoEdit]() {
      @Subscribe def handle(event: RoleEvent.GotoEdit) {
        val presenter = PresenterResolver.getPresenter(classOf[UserPermissionManagementPresenter])
        val role = event.getData.asInstanceOf[Role]
        presenter.go(container, new RoleScreenData.Edit(role))
      }
    })
    this.register(new ApplicationEventListener[RoleEvent.GotoRead]() {
      @Subscribe def handle(event: RoleEvent.GotoRead) {
        val presenter = PresenterResolver.getPresenter(classOf[UserPermissionManagementPresenter])
        presenter.go(container, new RoleScreenData.Read(event.getData.asInstanceOf[Integer]))
      }
    })
    this.register(new ApplicationEventListener[RoleEvent.GotoList]() {
      @Subscribe def handle(event: RoleEvent.GotoList) {
        val presenter = PresenterResolver.getPresenter(classOf[UserPermissionManagementPresenter])
        val criteria = new RoleSearchCriteria
        criteria.setSaccountid(new NumberSearchField(AppUI.getAccountId))
        presenter.go(container, new RoleScreenData.Search(criteria))
      }
    })
  }

  private def bingSettingEvents(): Unit = {
    this.register(new ApplicationEventListener[SettingEvent.GotoGeneralSetting]() {
      @Subscribe def handle(event: SettingEvent.GotoGeneralSetting) {
        val presenter = PresenterResolver.getPresenter(classOf[AccountSettingPresenter])
        presenter.go(container, new GeneralSetting())
      }
    })

    this.register(new ApplicationEventListener[SettingEvent.GotoTheme]() {
      @Subscribe def handle(event: SettingEvent.GotoTheme) {
        val presenter = PresenterResolver.getPresenter(classOf[AccountSettingPresenter])
        presenter.go(container, new ThemeCustomize())
      }
    })
  }

  private def bindSetupEvents(): Unit = {
    this.register(new ApplicationEventListener[SetupEvent.GotoSetupPage]() {
      @Subscribe def handle(event: SetupEvent.GotoSetupPage) {
        val presenter = PresenterResolver.getPresenter(classOf[SetupPresenter])
        presenter.go(container, null)
      }
    })
  }
}